changefeedccl: restore fine grained checkpoint from job proto

wenyihu6 opened the pull request:
As part of the fine-grained checkpointing project, changefeed now saves resolved
spans for each span above the high-water mark. This patch finalizes the
restoration process for fine-grained checkpoints. Upon resuming, changefeed
reloads span level checkpoints from the job proto and restores progress during
startup.
Restoration is required in three key areas:
1. When aggregators are initialized, including the initialization of each
aggregator frontier.
2. When `rangeFeedResumeFrontier` is initialized during every kvfeed.run.
3. When kvfeed determines which checkpointed spans to filter out from those
requiring backfill.

Note: This remains unused for now until issue https://github.com/cockroachdb/cockroach/issues/137692 is resolved.
NB: `reconcileJobStateWithLocalState` is intentionally left unchanged. We
decided to update it as part of issue https://github.com/cockroachdb/cockroach/issues/137692, as it depends on
`cachedState.SetCheckpoint`.

Resolves: https://github.com/cockroachdb/cockroach/issues/137693
Release Note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140379)
<!-- Reviewable:end -->

wenyihu6 committed:
changefeedccl: restore fine grained checkpoint from job proto

As part of the fine-grained checkpointing project, changefeed now saves resolved
spans for each span above the high-water mark. This patch finalizes the
restoration process for fine-grained checkpoints. Upon resuming, changefeed
reloads span level checkpoints from the job proto and restores progress during
startup.
Restoration is required in three key areas:
1. When aggregators are initialized, including the initialization of each
aggregator frontier.
2. When `rangeFeedResumeFrontier` is initialized during every kvfeed.run.
3. When kvfeed determines which checkpointed spans to filter out from those
requiring backfill.

Note: This remains unused for now until issue #137692 is resolved.
NB: `reconcileJobStateWithLocalState` is intentionally left unchanged. We
decided to update it as part of issue #137692, as it depends on
`cachedState.SetCheckpoint`.

Resolves: #137693
Release Note: None

aerfrei commented on pkg/ccl/changefeedccl/kvfeed/kv_feed.go:
```diff
@@ -592,11 +620,9 @@ func (f *kvFeed) runUntilTableEvent(ctx context.Context, resumeFrontier span.Fro
-			if _, err := resumeFrontier.Forward(s, f.checkpointTimestamp); err != nil {
-				return err
-			}
+	if f.spanLevelCheckpoint != nil || startFrom.Less(f.checkpointTimestamp) {
```
Could you leave a comment here saying the stuff you told me? a) What was the case we were doing the startFrom.Less check for? and b) why isn't it needed when f.spanLevelCheckpoint != nil?

wenyihu6 replied:
Done.

asg0451 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1777,6 +1777,7 @@ func (cf *changeFrontier) checkpointJobProgress(
 	}
 
 	cf.localState.SetHighwater(frontier)
+	// TODO(#137692)
```
so right now we're writing the new checkpoint .. to the local state only. which takes effect during transient errors or during node shutdown (not implemented yet). is that right?

wenyihu6 replied:
I still don't fully understand how this flow works. This PR is not touching this part of the flow and we are relying on [#137692](https://github.com/cockroachdb/cockroach/issues/137692#issuecomment-2637018153) to figure out how fine grained checkpointing work would work with `SetCheckpoint`. 

I think you are right though that frontier only saves to local state during transient errors but not node shutdown. But aggregators save their aggregator frontier during node shutdown https://github.com/cockroachdb/cockroach/blob/7c3fe6a353ea741161960ff6ebc98459a86246df/pkg/ccl/changefeedccl/changefeed_processors.go#L793-L799.

asg0451 commented on pkg/ccl/changefeedccl/kvfeed/kv_feed.go:
```diff
@@ -472,6 +478,26 @@ func filterCheckpointSpans(spans []roachpb.Span, completed []roachpb.Span) []roa
+
+	var spans []roachpb.Span
+	for _, sp := range newCheckpoint.Entries {
+		spans = append(spans, sp.Spans...)
```
why is this impl not the same as filterCheckpointSpans, with a spangroup and sub? they contain the same info, no?

wenyihu6 replied:
Good idea, I refactored this part of the code to reuse `filterCheckpointSpans`. Let me know your thoughts. Unless I'm missing something this impl is slightly different than filterCheckpointSpans because `TimestampSpansMap` is a map of (timestamp -> []Span). We need to extract the spans and flatten into a single slice. Does that match your understanding?

asg0451 commented on pkg/ccl/changefeedccl/kvfeed/kv_feed.go:
```diff
@@ -592,11 +618,15 @@ func (f *kvFeed) runUntilTableEvent(ctx context.Context, resumeFrontier span.Fro
-			if _, err := resumeFrontier.Forward(s, f.checkpointTimestamp); err != nil {
-				return err
-			}
+	if f.spanLevelCheckpoint != nil || startFrom.Less(f.checkpointTimestamp) {
```
perhaps instead of having `if new checkpoint not nil` checks everywhere, you could coerce the old checkpoint into the new format if the new one isnt set, and use only new-style logic afterwards everywhere in the code.

like `if spanlevelcheckpoint == nil, spanlevelcheckpoint = map{oldcheckpointts: [oldchkpoint_spans...]`

thoughts?

wenyihu6 replied:
Reached out to you on slack regarding this point. Not sure if I understand the benefits of this proposal.

asg0451 commented on pkg/ccl/changefeedccl/kvfeed/kv_feed.go:
```diff
@@ -592,11 +618,15 @@ func (f *kvFeed) runUntilTableEvent(ctx context.Context, resumeFrontier span.Fro
-			if _, err := resumeFrontier.Forward(s, f.checkpointTimestamp); err != nil {
-				return err
-			}
+	if f.spanLevelCheckpoint != nil || startFrom.Less(f.checkpointTimestamp) {
```
can you add an issue to remove this conditional check, since frontier.Forward cannot regress. and add a TODO(#issue) here

wenyihu6 replied:
Done.

asg0451 commented:
i'd still like to see a unification of the old checkpoint and the new checkpoint formats in the way we handle reading it in the code. if we transform {ts, []span} -> map{ts: []span} as soon as the changeagg/frontier starts up, it will simplify the code in a good way, and make writing the new version of the checkpoint free. @andyyang890 @wenyihu6 let's chat about this next week

wenyihu6 commented:
> i'd still like to see a unification of the old checkpoint and the new checkpoint formats in the way we handle reading it in the code. if we transform {ts, []span} -> map{ts: []span} as soon as the changeagg/frontier starts up, it will simplify the code in a good way, and make writing the new version of the checkpoint free. @andyyang890 @wenyihu6 let's chat about this next week

We chatted more over slack. I like this idea, but I'm also worried about possible regression. The current way makes sure that there are no regressions by passing in the exact same checkpointed ts and checkpointed spans as the old checkpoint code.

For example, with the new proposal, we would need to override the key if an empty key exists in the map here:
https://github.com/cockroachdb/cockroach/blob/bcfd86db76cbf42ce57dcee2cfe55e6497f6570a/pkg/ccl/changefeedccl/changefeed_processors.go#L613-L619. And there could be more examples that I haven't discovered yet. If we feel strongly about this way, we should address https://github.com/cockroachdb/cockroach/pull/140419 first. Which I think we would have to address in https://github.com/cockroachdb/cockroach/issues/137942 and https://github.com/cockroachdb/cockroach/issues/140509 as part of the fine grained checkpointing project. Also, this part of the code will be removed.

wenyihu6 commented:
TFTR!

bors r=aerfrei, asg0451
